Update Adobe Middleware

briemcnally opened the pull request:
**What does this PR do?**

- [x] Adds ID and Sync calls to the constructor of the middleware 
* Note it will `getMarketingCloudId` and then `syncAdvertisingId` if we don't have a `cachedMarketingCloudId`. Otherwise it will just `syncAdvertisingId` with the `cachedMarketingCloudId`. 
- [x]  Stores the `advertisingID` and `marketingCloudId` in iOS local storage (NSUserDefaults)
- [x] De-couples the API call into two API calls `getMarketingCloudId` and `syncAdvertsingId` 
- [x] Implements exponential backoff algorithm on API calls. The API calls will continue to retry if the call returns an error or if the responseStatusCode != 200. It will retry up to 10 times or until 5 minutes has hit. 
- [x] Improved error handling and incorporated into exponential backoff algorithm. An error will not be returned until the retry logic has exited. 
- [x] Is non-blocking. If there is no `cachedMarketingCloudId` events will continue to be sent to Segment without modification to payloads

f2prateek commented on analytics-ios-mcvid/Classes/SEGMCVIDTracker.m:
```diff
@@ -32,10 +37,170 @@ -(id)initWithOrganizationId:(NSString *)organizationId region:(NSString *)region
+            completion(marketingCloudId, nil);
+        } else {
+            dispatch_time_t delay = dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC * milliSecondsToWait);
+            self.backgroundQueue = dispatch_queue_create("com.exmaple.MyCustomQueue", NULL);
```
we should give it a name (e.g. `com.segment.mcvid`)

f2prateek replied:
I think we should create this queue once in initialization, and re-use it (as it can be expensive to create).

xavierjurado commented on analytics-ios-mcvid/Classes/SEGMCVIDTracker.m:
```diff
@@ -32,10 +37,169 @@ -(id)initWithOrganizationId:(NSString *)organizationId region:(NSString *)region
+    if (cachedMarketingCloudId.length == 0) {
+        [self getMarketingCloudId:organizationId maxRetryCount:maxRetryCount currentRetryCount:currentRetryCount maxRetryTimeSecs:maxRetryTimeSecs completion:^(NSString *marketingCloudId, NSError *error) {
+          [defaults setObject:marketingCloudId forKey:@"MarketingCloudId"];
+          [self syncMarketingCloudId:cachedAdvertisingId organizationId:organizationId marketingCloudId:cachedMarketingCloudId maxRetryCount:maxRetryCount currentRetryCount:currentRetryCount maxRetryTimeSecs:maxRetryTimeSecs completion:^(NSError *error) {
```
🚨 There's another bug here: `cachedMarketingCloudId` will always be `nil` at this point. Due to this bug and the one a few lines above both the marketing cloud id and the advertising id will be `nil`, thus making the id-sync request invalid.

xavierjurado commented on analytics-ios-mcvid/Classes/SEGMCVIDTracker.m:
```diff
@@ -32,10 +37,169 @@ -(id)initWithOrganizationId:(NSString *)organizationId region:(NSString *)region
+    }] resume];
+}
+
+- (NSURL *)createURL:advertisingId organizationId:(NSString *)organizationId marketingCloudId:(NSString *)marketingCloudId {
```
💡 Taking into account the changes we need in order to make our own id-sync calls, you will probably want to change this method into something like:

``` suggestion
- (NSURL *)URLFromOrganizationId:(NSString *)organizationId marketingCloudId:(NSString *)marketingCloudId IntegrationCode:(NSString *)integrationCode userIdentifier:(NSString *)userIdentifier {
```

The last two parameters would be optional and be used only by your idsync call, passing `DSID_20915` as the `integrationCode` and the advertising identifier as the `userIdentifier`. Inside this method you would then concatenate them using the `%01` separator just like you do with your existing `deviceTypeKey` and `advertisingId`.

xavierjurado commented on analytics-ios-mcvid/Classes/SEGMCVIDTracker.m:
```diff
@@ -61,114 +225,52 @@ - (void)sendRequestAdobeExperienceCloud:(NSString *)advertisingId organizationId
+        [queryItems addObject:[NSURLQueryItem queryItemWithName:marketingCloudIdKey value:marketingCloudId]];
+        NSString *encodedAdvertisingValue = [NSString stringWithFormat:@"%@%@%@", deviceTypeKey, separator, advertisingId];
+        //removes %25 html encoding of '%'
+        NSString *normalAdvertisingValue = [encodedAdvertisingValue stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
```
🚨 Another bug here: by invoking `stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding` you are effectively removing the `%01` separator from `encodedAdvertisingValue`, making this value and the id-sync request invalid.

briemcnally commented:
@xavierjurado I have addressed a lot of the comments/issues with the iOS middleware. I have added a list of the addressed issues:

- [ ]  Remove `getMarketingCloudId` method from header file 
- [ ] Remove pointers to NSUInteger properties and remove them from public interface in header file 
- [ ] Remove `staticMarketingCloudId` property from header file
- [ ] Remove `stringByRemovingPercentEncoding` property from header file
- [ ] Remove `cachedAdvertisingId` property from header file
- [ ] Remove `staticMarketingCloudId` property from header file
- [ ] Add `cachedAdvertisingId` and `cachedMarketingCloudId` propertied to the interface in the `.m`. Remove it the parameter from the method signatures. 
- [ ] Change `syncMarketingCloudId` to fit your use case for performing your own idsync calls 
- [ ] Update the `createURL` function accordingly given the above changes
- [ ] Add  nullability annotations 
- [ ] Fix initialization for clean install where `cachedAdvertisingId` is nil. Now it is updated with value from `SEGIDFA()` 
- [ ] Added prefixes to NSUserDefault keys to avoid collisions  (both `advertisingId` and `marketingCloudId`)
- [ ] Updated initialization to match functionality of android where sync call is only triggered is if the `advertisingID` changes. 
- [ ] Exported the MCVIDAdobeErrorKey
- [ ] Fixed exponential backoff logic to `t = 2 ^ retryCount` and fixed variable name to `secondsToWait`
- [ ] Removed logic surrounding `millisecondsWaited`
- [ ] Removed check for `organizationId` to continue 
- [ ] Checked if the payload contains "Adobe Analytics" integration options and merged the dictionaries
- [ ] Fixed `cachedMarketingCloudId` being nil on initialization on clean install causing the id-sync to fail

xavierjurado commented on analytics-ios-mcvid/Classes/SEGMCVIDTracker.m:
```diff
@@ -19,204 +25,348 @@ - (id)initWithCode:(MCVIDAdobeErrorCode)code message:(NSString *)message error:(
+            [existingIntegrations removeObjectForKey:@"Adobe Analytics"];
+            [integrations addEntriesFromDictionary:existingIntegrations];
+            [integrations setObject:adobeOptions forKey:@"Adobe Analytics"];
+        }
```
🚨 We have another serious bug here: if the payload's integrations dictionary is `nil` we won't send the `marketingCloudVisitorId`. The same applies to all the other types of payloads below.

xavierjurado replied:
Also please, let's try to move this code into a single place instead of copying it into every kind of payload. I also feel like we could simplify this code a lot.
